pkg/endpointslice/map.go

```diff
@@ -0,0 +1,70 @@
+package endpointslice
+
+import (
+	"sync"
+
+	"github.com/submariner-io/lighthouse/pkg/constants"
+	discovery "k8s.io/api/discovery/v1beta1"
+)
+
+type endpointInfo struct {
+	key     string
+	hostIPs map[string][]string
+}
+
+type Map struct {
+	epMap map[string]*endpointInfo
+	sync.RWMutex
+}
+
+func (m *Map) GetIPs(hostname, cluster, namespace, name string) ([]string, bool) {
+	key := keyFunc(namespace, name, cluster)
+
+	result, ok := m.epMap[key]
+	if !ok {
+		return nil, false
+	}
+
+	ips, ok := result.hostIPs[hostname]
+
+	return ips, ok
+}
+
+func NewMap() *Map {
+	return &Map{
+		epMap: make(map[string]*endpointInfo),
+	}
+}
+
+func (m *Map) Put(es *discovery.EndpointSlice) {
+	key, ok := es.Labels[constants.LabelServiceImportName]
+	if !ok {
+		return
+	}
+
+	m.Lock()
+	defer m.Unlock()
+
+	epInfo := &endpointInfo{
+		key:     key,
+		hostIPs: make(map[string][]string),
+	}
+
+	for _, endpoint := range es.Endpoints {
+		epInfo.hostIPs[*endpoint.Hostname] = endpoint.Addresses
```

Comment:
We should check if `Hostname` is nil as it is optional. Also not sure if `Addresses` could be nil but wouldn't hurt to check as well.

Reply:
Fixed in subsequent commit.